contracts/utils/Strings.sol

```diff
@@ -113,4 +128,261 @@ library Strings {
+        bytes memory buffer = bytes(input);
+
+        // check presence of a negative sign.
+        bool positiveSign = bytes1(buffer.unsafeReadBytesOffset(begin)) == bytes1("+");
+        bool negativeSign = bytes1(buffer.unsafeReadBytesOffset(begin)) == bytes1("-");
+        uint256 offset = (positiveSign || negativeSign).toUint();
+        int8 factor = negativeSign ? int8(-1) : int8(1);
+
+        int256 result = 0;
+        for (uint256 i = begin + offset; i < end; ++i) {
+            uint8 chr = _tryParseChr(buffer[i]);
+            if (chr > 9) return (false, 0);
+            result *= 10;
+            result += factor * int8(chr);
+        }
+        return (true, result);
+    }
+
+    /**
+     * @dev Parse a hexadecimal string (with or without "0x" prefix), and returns the value as a `uint256`.
+     *
+     * Requirements:
+     * - The string must be formatted as `(0x)?[0-9a-fA-F]*`
+     * - The result must fit in an `uint256` type.
+     */
+    function parseHex(string memory input) internal pure returns (uint256) {
+        return parseHex(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseHex} that parses a substring of `input` located between position `begin` (included) and
+     * `end` (excluded).
+     *
+     * Requirements:
+     * - The substring must be formatted as `(0x)?[0-9a-fA-F]*`
+     * - The result must fit in an `uint256` type.
+     */
+    function parseHex(string memory input, uint256 begin, uint256 end) internal pure returns (uint256) {
+        (bool success, uint256 value) = tryParseHex(input, begin, end);
+        if (!success) revert StringsInvalidChar();
+        return value;
+    }
+
+    /**
+     * @dev Variant of {parseHex-string} that returns false if the parsing fails because of an invalid character.
+     *
+     * Requirements:
+     * - The result must fit into an `uint256` type.
+     */
+    function tryParseHex(string memory input) internal pure returns (bool success, uint256 value) {
+        return tryParseHex(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseHex-string-uint256-uint256} that returns false if the parsing fails because of an
+     * invalid character.
+     *
+     * Requirements:
+     * - The result must fit into an `uint256` type.
+     */
+    function tryParseHex(
+        string memory input,
+        uint256 begin,
+        uint256 end
+    ) internal pure returns (bool success, uint256 value) {
+        bytes memory buffer = bytes(input);
+
+        // skip 0x prefix if present
+        bool hasPrefix = bytes2(buffer.unsafeReadBytesOffset(begin)) == bytes2("0x");
+        uint256 offset = hasPrefix.toUint() * 2;
+
+        uint256 result = 0;
+        for (uint256 i = begin + offset; i < end; ++i) {
+            uint8 chr = _tryParseChr(buffer[i]);
+            if (chr > 15) return (false, 0);
+            result *= 16;
+            result += chr;
```

Comment:
```suggestion
            assembly ("memory-safe") {
                result := or(shl(4, result), chr)
            }
```

Would something like this be more efficient? Why don't we return a `bytes32`.

Reply:
If the result overflow, this will return incorrect values. Also its way less readable

If you want to save gas, you could do just the same logic using unchecked instead of assembly ... but again you would have silent invalid result in case of overflow.

Reply:
Overflow occurs if the length is > 64 (assuming no leading 0s) so it’s easy to calculate and revert.

Reply:
Not sure we should assume there are no leading zero.
Also, that won't work for the base 10 variant.

Reply:
What you committed makes sense--I prefer the shift as its probably more logical when thinking of bytes32 hex. Still confused why a `uint256` is returned.

Reply:
I think returning uint256 of bytes32 is equivalent. 
@ernestognw do you have a preference ?

Reply:
Agree is equivalent. I personally prefer `uint256` since parsing string to bytes is already easy, and I seems that the only reason to parse to a `bytes32` would be to use it as a number (otherwise it'd be already `bytes32`)

Reply:
> What you committed makes sense--I prefer the shift as its probably more logical when thinking of bytes32 hex.

Problem with the shift is it gives invalid behavior when result does not fit in 256 bits :/ (no overflow check)

Reply:
> > What you committed makes sense--I prefer the shift as its probably more logical when thinking of bytes32 hex.
> 
> Problem with the shift is it gives invalid behavior when result does not fit in 256 bits :/ (no overflow check)

Checked shift as implemented is good.